### ISSUE TITLE
[Feaure Request] implement a way to register a callback to be called when class is instantiated

### ISSUE THREAD

**Issue by blikblum** [zheksoon/dioma#5]
Bottlejs has a concept of decorator https://github.com/young-steveo/bottlejs?tab=readme-ov-file#decorators that is a function that runs when class / service is instantiated 

Currently i use to register event handlers when a class declares one or more

**Comment by zheksoon** [comment]
Thanks for the request, totally makes sense :)

**Comment by zheksoon** [comment]
Done, see https://github.com/zheksoon/dioma?tab=readme-ov-file#injection-hooks section of README

**Comment by blikblum** [comment]
Many thanks.

Can you show me how can i get the instance and or the class using `container`, `descriptor`, `args` parameters?

**Comment by zheksoon** [comment]
Sure! The descriptor is the object that was used in `.register` method. Example from README:
```javascript
wild.register({ token: animalToken, class: Dog, scope: Scopes.Transient() });
```
So you can use `descriptor.class`, `descriptor.token`, `descriptor.scope` from it, if specified when registering the class.
As the `beforeCreate` is called before the class instance is available, there is no way to get it. Maybe to add another hook like `afterCreate`, what do you think?

**Comment by blikblum** [comment]
> Maybe to add another hook like `afterCreate`, what do you think?

Yes. To cover the use case i cited and covered by bottlejs#decorator.

If i understand correct, `beforeInject` is called every time is injected right? (just for understanding)

Some notes:
 - I cannot see a use case for `beforeCreate` since AFAIK is not possible to change how instance is created unless modifying descriptor which is something i would not recommend thinking of API design
 - Initially i think in a way to register the hook in the container instance being called for all classes registered in it (like in bottlejs). But is something that can be easily overcome by creating a helper function in "userland" that registers a class using a predefined hook. So i would not pursue any change here

**Comment by blikblum** [comment]
Here's my concrete user case:

Declare the class to be injected
```js
// @service register a class in the DI container
// @taskHandler register the method in the a property of class constructor to be used as event handlers
@service('schedulesStore', ['appStore'])
class SchedulesStore extends SchedulesQuery {
  constructor(appStore) {
    super()
    bindRefParams(appStore, this, {
      customerId: 'currentCustomer',
    })
    this.observe()
  }

  @taskHandler('save-schedule')
  saveSchedule({model, attrs, dirtyAttrs}) {
    if(model.isNew()) {
      return this.create(attrs)
    }

    return model.save(dirtyAttrs, { patch: true })
  }
}

export { SchedulesStore }
```

start listening to the events

```js
// in the start of app
// bindClassTasks checks if there are handlers registered in constructor and configure the event listeners
registry.decorator((instance) => {
  bindClassTasks(instance)
  return instance
})
```

**Comment by zheksoon** [comment]
Ah, ok, I see. Will think about the implementation. But PRs are welcome! :)
